### PR TITLE
FOUNDATIONS: Add Post Impression

### PR DESCRIPTION
**Pull request by evangelinedrink**

### CODE REVIEW

**Comment by hassanhabib** [comment]
@ElbekDeveloper take a look at this PR please.

File: Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Exceptions.Add.cs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                    expectedPostImpressionDependencyException))),
+                        Times.Once);
+
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async void ShouldThrowValidationExceptionOnAddIfReferenceErrorOccursAndLogItAsync()
```
Comment (ElbekDeveloper) [inline Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Exceptions.Add.cs:157]: I would call it `ShouldThrowDependencyValidationExceptionOnAddIfReferenceErrorOccursAndLogItAsync`

File: Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Logic.Add.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+namespace Taarafo.Core.Tests.Unit.Services.Foundations.PostImpressions
+{
+    public partial class PostImpressionServiceTests
+    {
+        [Fact]
+        public async Task ShouldAddPostImpressionAsync()
+        {
+            //given
+            DateTimeOffset dateTime = GetRandomDateTimeOffset();
```
Comment (ElbekDeveloper) [inline Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Logic.Add.cs:22]: call it `randomDateTime`

File: Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+            await Assert.ThrowsAsync<PostImpressionValidationException>(() =>
+                addPostImpressionTask.AsTask());
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedPostImpressionValidationException))),
+                        Times.Once);
+
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
```
Comment (ElbekDeveloper) [inline Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs:43]: Verify no other calls from `dateTimeBroker`  as well

File: Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedPostImpressionValidationException))),
+                        Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.InsertPostImpressionAsync(invalidPostImpression),
+                    Times.Never);
+
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
```
Comment (ElbekDeveloper) [inline Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs:106]: Do the same for rest

File: Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+            var invalidPostImpressionException =
+                new InvalidPostImpressionException();
+
+            invalidPostImpressionException.AddData(
+                key: nameof(PostImpression.PostId),
+                values: "Id is required.");
+
+            invalidPostImpressionException.AddData(
+                key: nameof(PostImpression.Post),
+                values: "Post is required.");
```
Comment (ElbekDeveloper) [inline Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs:67]: Message should be as `Object is required`

File: Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+            invalidPostImpressionException.AddData(
+                key: nameof(PostImpression.ProfileId),
+                values: "Id is required.");
+
+            invalidPostImpressionException.AddData(
+                key: nameof(PostImpression.Profile),
+                values: "Profile is required.");
```
Comment (ElbekDeveloper) [inline Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs:75]: Message should be as `Object is required`

File: Taarafo.Core/Models/PostImpressions/Exceptions/NullPostImpressionException.cs [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+// ---------------------------------------------------------------
+
+using Xeptions;
+
+namespace Taarafo.Core.Models.PostImpressions.Exceptions
+{
+    public class NullPostImpressionException : Xeption
+    {
+        public NullPostImpressionException()
+            : base(message: "Post Impression is null.")
```
Comment (ElbekDeveloper) [inline Taarafo.Core/Models/PostImpressions/Exceptions/NullPostImpressionException.cs:13]: Some messages have `Post Impression` others have `Post impression`. Please, update messages to be consistent. I like `Post impression`...

File: Taarafo.Core/Services/Foundations/PostImpressions/PostImpressionService.Validations.cs [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        {
+            Condition = date == default,
+            Message = "Date is required."
+        };
+
+        private static dynamic IsInvalid(Post post) => new
+        {
+            Condition = post == null,
+            Message = "Post is required."
+        };
```
Comment (ElbekDeveloper) [inline Taarafo.Core/Services/Foundations/PostImpressions/PostImpressionService.Validations.cs:61]:
```
private static dynamic IsInvalid(object @object) => new
{
    Condition = @object  is null,
    Message = "Object is required."
};
```

**Review by ElbekDeveloper — CHANGES_REQUESTED** [review]

**Comment by ElbekDeveloper** [comment]
@hassanhabib  
image
The beginning of the PR has wrong one commit. I left what can be changed. But, I want you to give the final decision on what to do with such commit history

File: Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Exceptions.Add.cs (hunk 1 above)
Comment (evangelinedrink) [reply Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Exceptions.Add.cs]: I have added the Dependency to the name of this test in the commit named CODERUB: Fixed Name of Test By Adding Dependency.

File: Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Logic.Add.cs (hunk 2 above)
Comment (evangelinedrink) [reply Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Logic.Add.cs]: I have renamed the variable in the commit named CODERUB: Changed Variable Name To randomDateTime

File: Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs (hunk 3 above)
Comment (evangelinedrink) [reply Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs:43]: I added calls from dateTimeBroker in the commit named CODERUB: Created The DateTimeBrokerMock For Tests In Validation

File: Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs (hunk 4 above)
Comment (evangelinedrink) [reply Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs:112]: I was able to add the dateTimeBroker for the other tests in the commit  CODERUB: Created The DateTimeBrokerMock For Tests In Validation

File: Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs (hunk 6 above)
Comment (evangelinedrink) [reply Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs:76]: Changed the message in the commit named CODERUB: Changed Message For Profile And Post

File: Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs (hunk 5 above)
Comment (evangelinedrink) [reply Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs:68]: Changed the message in the commit named CODERUB: Changed Message For Profile And Post

File: Taarafo.Core/Models/PostImpressions/Exceptions/NullPostImpressionException.cs (hunk 7 above)
Comment (evangelinedrink) [reply Taarafo.Core/Models/PostImpressions/Exceptions/NullPostImpressionException.cs:13]: I updated the messages in CODERUB: Updated Messages For Post Impression Exceptions. Pull Request link is here: https://github.com/hassanhabib/Taarafo.Core/pull/221

File: Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs (hunk 5 above)
Comment (evangelinedrink) [reply Taarafo.Core.Tests.Unit/Services/Foundations/PostImpressions/PostImpressionServiceTests.Validations.Add.cs:68]: The Pull Request with the changes can be found here: https://github.com/hassanhabib/Taarafo.Core/pull/221

File: Taarafo.Core/Services/Foundations/PostImpressions/PostImpressionService.Validations.cs (hunk 8 above)
Comment (evangelinedrink) [reply Taarafo.Core/Services/Foundations/PostImpressions/PostImpressionService.Validations.cs:61]: I was able to change this, but one of the validation test is now failing. The test starting in line 164 in the PostImpressionServiceTests.Validations.Add.cs file is stating that no exception is being thrown, but it is expecting the PostImpressionValidationException.

Comment (evangelinedrink) [reply Taarafo.Core/Services/Foundations/PostImpressions/PostImpressionService.Validations.cs:61]: It is working now, please ignore my last message.  I made sure to remove the periods on all of the validation messages. The updated section is in the commit named CODERUB: Profile And Post Replaced with Object. The PR link: https://github.com/hassanhabib/Taarafo.Core/pull/221

**Pull request merged**